sui_core/src/safe_client.rs

```diff
@@ -310,39 +313,61 @@ where
-                .then(move |batch_info_item| {
-                    let req_clone = request.clone();
-                    let client = client.clone();
-                    async move {
-                        match &batch_info_item {
-                            Ok(BatchInfoResponseItem(UpdateItem::Batch(signed_batch))) => {
-                                if let Err(err) =
-                                    client.check_update_item_batch_response(req_clone, signed_batch)
-                                {
-                                    client.report_client_error(err.clone());
-                                    return Err(err);
-                                }
-                                batch_info_item
-                            }
-                            Ok(BatchInfoResponseItem(UpdateItem::Transaction((_seq, _digest)))) => {
-                                batch_info_item
-                            }
-                            Err(e) => Err(e.clone()),
+        let address = self.address;
+        let stream = Box::pin(batch_info_items.scan(
+            (0u64, None),
+            move |(seq, txs_and_last_batch), batch_info_item| {
+                let req_clone = request.clone();
+                let client = client.clone();
+
+                // We check if we have exceeded the batch boundary for this request.
+                if !(*seq < request.end) {
+                    // If we exceed it return None to end stream
+                    return futures::future::ready(None);
+                }
+
+                let x = match &batch_info_item {
+                    Ok(BatchInfoResponseItem(UpdateItem::Batch(signed_batch))) => {
+                        println!("batch {:?}", signed_batch.batch);
+                        if let Err(err) = client.check_update_item_batch_response(
+                            req_clone,
+                            &signed_batch,
+                            &txs_and_last_batch,
+                        ) {
+                            client.report_client_error(err.clone());
+                            Some(Err(err))
+                        } else {
+                            // Save the seqeunce number of this batch
+                            *seq = signed_batch.batch.next_sequence_number;
+                            // Insert a fresh vector for the new batch of transactions
+                            let _ =
+                                txs_and_last_batch.insert((Vec::new(), signed_batch.batch.clone()));
+                            Some(batch_info_item)
                         }
                     }
-                })
-                .take(seq_to_be_returned),
-        );
+                    Ok(BatchInfoResponseItem(UpdateItem::Transaction((seq, digest)))) => {
+                        // A stream always starts with a batch, so the previous should have initialized it.
+                        // And here we insert the tuple into the batch.
```

Comment:
In one of the batch tests, I saw that there were a few individual transactions sent before a batch was sent, also is it not ever expected that we would have individual transactions sent only like how we see in the benchmark with the default parameters set?

Reply:
Hm, right now we indeed have a gap in testing. We test only against a mock server. It would be nice to test against a real authority to make sure we catch something like this.

Reply:
As I remembered the follower server side should be sending batch --> (trasnactions+ -> batch) + , so all transactions are fully enclosed by batches. We should test to see if there is any issue with this?